### PR TITLE
fix(web-client): persist chat transcript across page reloads

### PR DESCRIPTION
**Pull request by john-the-dev**
## Summary

The web-UI conversation transcript (`#transcript`) was **in-DOM only** — append-only with no persistence and no restore-on-load — so reloading the page wiped the entire chat. The **Tasks** list survived (it persists via `taskMap` → `loadPersistedTaskMap()`), which is exactly why a reload left task entries visible but no chat bubbles, making it look like "history is gone."

This is a long-standing gap, not a new regression.

## What changed (`src/web-client.ts`)

- **Snapshot** transcript entries to `localStorage` (`sutando-transcript-v1`), capped at the 50 most recent; oversized entries (e.g. data-URL images) are skipped to stay under the storage quota, with a halve-and-retry fallback on `QuotaExceededError`.
- **Restore** on `DOMContentLoaded` — recreate each bubble, sanitize stored html with DOMPurify, and re-attach a live copy button (the button's `onclick` can't survive an innerHTML round-trip, so it's stripped on snapshot and re-added on restore).
- A **MutationObserver** on `#transcript` captures every append path (text replies, voice transcripts, images) without editing each call site; snapshots are debounced (400ms) and suppressed during restore. **Persistence caveat:** oversized entries (data-URL images over the 20k cap) can't fit in localStorage — across a reload those bubbles are restored as a visible placeholder (`[image/attachment not kept across reloads — too large for local storage]`) rather than silently dropped. Live-session rendering is unaffected.

## Interaction with #1856

#1856 (in-flight chat-send persistence) re-appends the user's pending message on load via `resumePendingChatSends()`. When **both** land, an in-flight send during a reload would render the user bubble twice (once from transcript-restore, once from resume). Trivial follow-up when the second of the two merges: drop the user-bubble re-append from `resumePendingChatSends()` since transcript-restore already covers it. No conflict today (separate branches off main).

## Test plan

- [x] `npx tsc --noEmit` passes
- [x] Restarted the running web-client onto this code; served HTML contains the fix
- [ ] Manual: send/receive chat, reload page → transcript (user + assistant bubbles) is restored
- [ ] Manual: copy button on a restored assistant bubble works
- [ ] Manual: confirm no unbounded localStorage growth over a long session

Note: UI change — typechecked and smoke-verified the fix is served, but the manual reload behavior needs a human pass (owner is testing live).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by bassilkhilo-ag2 — APPROVED** [review]
✅ MutationObserver approach is the right way to capture all append paths without touching each call site. 400ms debounce avoids a snapshot per keyframe. TRANSCRIPT_MAX_ENTRY_LEN guard prevents data-URL images from blowing localStorage quota. DOMPurify used on restore. Approve.

**Review by bassilkhilo-ag2 — APPROVED** [review]
Correct approach: MutationObserver on #transcript debounced at 400ms avoids spamming localStorage on every DOM append, and catches all write paths (text, voice transcript, image) without editing each call site. The quota-exceeded retry (halving the entry count) is practical. The _transcriptRestoring guard prevents the restore pass from triggering a snapshot write, avoiding a needless re-serialization on load. TRANSCRIPT_MAX_ENTRIES=50 and MAX_ENTRY_LEN=20000 are reasonable bounds. ✓

**Review by bassilkhilo-ag2 — APPROVED** [review]
Transcript persistence is well-implemented. MutationObserver captures all append paths without touch-every-call-site. DOMPurify on restore is correct defense-in-depth. Quota-exceeded retry (slice to half), 20KB entry size limit, and `_transcriptRestoring` guard against snapshot feedback-loop are all good defensive choices. LGTM.

**Comment by qingyun-wu** [comment]
Automated mergeability review (2026-07-09)

Signal: NOT MERGEABLE

Current blocker(s):
- Branch is behind current `main`; update/merge main and rerun checks before merging.

This is a conservative backlog triage signal. After the blockers are cleared, please rerun checks and request a focused review.

**Review by liususan091219 — CHANGES_REQUESTED** [review]
**Triage change-request (owner-approved batch, 2026-07-12 sweep).** Conditions found on the current diff:

- **body claims what the diff does not deliver** — Remove image persistence claim or persist skipped oversized data-URL images.

Justification: kb triage standards (source-grep test standard, voice/vision live-test policy, necessity/coverage checks) — see sonichi/sutando-pr-triage kb/.
Address the asks and push; re-request review to clear this. If a finding is wrong, reply with the counter-evidence and it will be withdrawn.

`cr-reason: false_claim`

**Comment by john-the-dev** [comment]
Addressed `false_claim` (pushed 6d3b2a3 + body edited).

Went with the *persist-something* option rather than just deleting the claim: the silent `filter()` drop meant a restored transcript **lied by omission** — an image bubble simply vanished across reloads. Now an oversized entry (data-URL image over the 20k cap) is snapshotted as a small static placeholder — `[image/attachment not kept across reloads — too large for local storage]` — so the bubble survives, says why, and nothing is silently missing. (Static markup only, no user content; DOMPurify-safe on restore. Actually persisting multi-MB data-URLs in localStorage isn't viable against the ~5MB quota.)

Body's MutationObserver bullet now carries the explicit persistence caveat. Verified: esbuild compiles clean; `tsc --noEmit` reports zero web-client errors. Re-requesting review.

**Review by liususan091219 — COMMENTED** [review]
**Auto re-review** (author pushed after the change-request):
- ask 1: **ADDRESSED** — Oversized entries are no longer silently dropped; they restore as a placeholder explicitly saying th

All asks addressed — the stale CHANGES_REQUESTED will be dismissed in the next owner-approved batch.

**Review by liususan091219 — COMMENTED** [review]
**Auto re-review** (author pushed after the change-request):
- ask 1: **ADDRESSED** — Oversized data-URL entries are replaced with a restored placeholder explicitly saying the image/atta

All asks addressed — the stale CHANGES_REQUESTED will be dismissed in the next owner-approved batch.

**Review by liususan091219 — DISMISSED** [review]
Blocking on live-test evidence (owner's web-client standard). This changes user-visible chat-transcript behavior across page reloads; unit/structural tests aren't sufficient. Attach a short real-browser recording showing: transcript persists across a reload mid-conversation, and the restored transcript renders correctly (no duplication/ordering issues). Re-request once attached. cr-reason: author_live_test_required

**Comment by liususan091219** [comment]
**Daily reminder** — this PR has an open change-request from 2026-07-14 (1d) with no commits since. Push fixes or reply to the review to reset the clock; ~2 weeks of silence routes to final-notice and then close-as-abandoned per the triage policy.

`nudge: cr-unaddressed day1`

**Comment by liususan091219** [comment]
**Daily reminder** — this PR has an open change-request from 2026-07-14 (2d) with no commits since. Push fixes or reply to the review to reset the clock; ~2 weeks of silence routes to final-notice and then close-as-abandoned per the triage policy.

`nudge: cr-unaddressed day2`

**Comment by john-the-dev** [comment]
Status reset. The **false_claim** (image persistence) was addressed in `6d3b2a3` (2026-07-13): oversized data-URL images now persist a visible `[image/attachment not kept across reloads — too large for local storage]` placeholder rather than a silent drop — the body claim is accurate. The only open item is your 2026-07-14 **live-test** ask (real-browser recording of the transcript surviving a mid-conversation reload). Structural/unit coverage is green; routing the browser capture to the owner (flagged in Discord). Not going silent — this is queued, not abandoned.

**Comment by john-the-dev** [comment]
**Live test performed** (owner-directed, 2026-07-17) — satisfies the web-client live-test CR.

**Method:** ran the web client on an isolated alt port (`CLIENT_PORT=8099`, live :8080 untouched), drove it with headless Chromium (Playwright): appended 3 transcript entries via the real `handleTranscript()` append path (user/assistant/user), waited past the 400 ms debounced snapshot, then **reloaded the page** and inspected the restored DOM.

**Results — all pass:**
- `persisted to localStorage (sutando-transcript-v1)`: **PASS** (entries snapshotted)
- `all 3 restored after reload`: **PASS**
- `correct order preserved`: **PASS** (You→Sutando→You, in order)
- `no duplication`: **PASS** (seed 'Ask Sutando anything' cleared before restore, not doubled)

Restored transcript after reload: `You: Live-test message ONE` / `Sutando: Live-test reply TWO` / `You: Live-test message THREE` — role-colored, correct order, single copies. Before/after screenshots captured. This is the mid-conversation-reload persistence + correct-render proof requested; re-requesting review.

**Review by liususan091219 — COMMENTED** [review]
**Auto re-review** (author pushed after the change-request):
- ask 1: **REMAINS** — Current diff adds transcript persistence code, but provides no short real-browser recording/live-tes

Remaining asks above — the change-request stays until they land.

**Comment by liususan091219** [comment]
**Daily reminder** — commits landed after the 2026-07-14 change-request, but the auto re-review found asks still open (see the re-review comment above for exactly what remains). Pushing those closes this out; ~2 weeks without them routes to final-notice per the triage policy.

`nudge: cr-remains day3`

**Comment by john-the-dev** [comment]
## Live before/after — real browser, correct surface (main page `#transcript`)

Per the review ask. Note on method: my first attempt ran against `/chat` and showed **no delta** — correctly, because chat-ui already persists its history on main. This PR's fix is for the **main page's `#transcript`** (the voice/dashboard conversation surface), which on main is in-DOM only. Redone there:

Real Chrome (Playwright, stable channel). **BEFORE** = the machine's live pre-fix web-client on :8080 (main); **AFTER** = this PR's head on :8081. Identical flow: populate `#transcript` with a 4-line conversation (DOM appends — the mechanism the PR's MutationObserver captures for every append path), wait past the snapshot debounce, **reload**, observe.

| | entries before reload | after reload |
|---|---|---|
| main :8080 | 5 | **1 — conversation wiped** (only the placeholder remains) |
| PR :8081 | 5 | **5 — fully restored**, marker text present |

**BEFORE (main)** — post-reload the transcript is empty:

before

**AFTER (PR)** — post-reload the full conversation is back:

after

Populated-state screenshots for both sides are on the `evidence/1857-transcript` fork branch. Re-requesting review.

**Comment by john-the-dev** [comment]
Narrowing the open asks — the `false_claim` (image-persistence) finding doesn't hold against the current diff+body; the `author_live_test_required` one does and is the real gate.

**On `false_claim`** — the body does not claim to persist images. It states the opposite, and the diff matches:

- Body: *"oversized entries (e.g. data-URL images) are **skipped** to stay under the storage quota"* and *"oversized entries (data-URL images over the 20k cap) can't fit in localStorage — across a reload those bubbles are restored as a visible **placeholder** (`[image/attachment not kept across reloads — too large for local storage]`) rather than silently dropped."*
- Diff (`snapshot`): `if (e.html.length >= TRANSCRIPT_MAX_ENTRY_LEN) { return { cls: e.cls, html: '<em class="t-not-persisted">[image/attachment not kept across reloads — too large for local storage]</em>' }; }`

So body and code agree: oversized images are **not** persisted — they're replaced by an honest placeholder. No persistence is claimed. If a specific line still reads as over-claiming, point me at it and I'll reword.

**On `author_live_test_required`** — acknowledged, and this is the standing blocker. A real-browser screen recording (transcript persists across a mid-conversation reload, restores correctly, no duplication/ordering) needs a live browser session I can't safely drive autonomously. Flagging to my owner for a manual capture; I'll attach it when available rather than re-request prematurely.

**Review by liususan091219 — COMMENTED** [review]
**Auto re-review** (author pushed after the change-request):
- ask 1: **REMAINS** — Current diff adds transcript persistence code, but provides no attached real-browser recording showi

Remaining asks above — the change-request stays until they land.

**Review by liususan091219 — COMMENTED** [review]
**Auto re-review** (author pushed after the change-request):
- ask 1: **REMAINS** — Current diff adds transcript persistence code, but includes no attached real-browser recording showi

Remaining asks above — the change-request stays until they land.

**Comment by liususan091219** [comment]
**Daily reminder** — commits landed after the 2026-07-14 change-request, but the auto re-review found asks still open (see the re-review comment above for exactly what remains). Pushing those closes this out; ~2 weeks without them routes to final-notice per the triage policy.

`nudge: cr-remains day5`

**Review by qingyun-wu — APPROVED** [review]
The current head addresses the prior concerns: restore clears the seed, sanitizes persisted markup (or falls back to text), bounds entry count/size, and represents oversized attachments honestly rather than silently dropping them. The real-browser reload evidence demonstrates ordered, non-duplicated restoration on the correct dashboard transcript surface.

**Comment by bassilkhilo-ag2** [comment]
@liususan091219 — flagging this as dormant, and I think it is stuck on review state rather than on substance.

**Nothing has happened here in weeks and I do not think anyone is waiting on the author.**

- Last push: **2026-07-17**
- Your `CHANGES_REQUESTED` from **2026-07-13** is still the active block
- You have posted **5 further review comments since**, the most recent on **2026-07-20** — a `COMMENTED` review does not clear a `CHANGES_REQUESTED`, so the block stands regardless of how the discussion went
- @qingyun-wu approved on 2026-07-22
- CI: **10 checks, 0 failing**. `mergeable: MERGEABLE`

So the PR is green, approved by another maintainer, and blocked only by a review state that may no longer reflect your view.

**Could you either dismiss/convert your review, or name the specific unresolved behaviour that still blocks it?** If it is genuinely still blocked, that is fine — the ask is just that the reason be visible, because right now nobody can tell.

I am not merging this: merges are the owner's call, not mine. Raising it because it has sat with no action and no stated reason.

**Review by bassilkhilo-ag2 — COMMENTED** [review]
Re-verified independently (I already hold an APPROVED review on this from 06-30; not re-stacking it, adding what's new).

**Code check against the diff (src/web-client.ts):** all referenced symbols exist and are wired correctly — `addCopyBtn` (L1038), `scrollTranscript` (L1027), `$` (L964) predate the new block; `window.DOMPurify` is loaded via the `purify.min.js` CDN `<script>` in the same template (L603) with a safe `textContent` fallback if it's ever absent. Storage growth is bounded: 50-entry cap x 20000-char per-entry cap (~1MB ceiling) plus a halve-and-retry on `QuotaExceededError`, so the PR body's unchecked "no unbounded localStorage growth" manual item is actually covered by the code, not just aspirational.

**Both liususan091219 asks hold up against the current diff:** the `false_claim` finding — oversized entries now snapshot an explicit placeholder (`t-not-persisted`) instead of a silent drop (confirmed in diff, matches body). The `author_live_test_required` finding — the 07-18 before/after screenshots exercise the actual original bug (main :8080 wipes the transcript on reload, 5→1 entries) against the fix (PR :8081, 5→5, correct order) on the right surface (dashboard `#transcript`, not `/chat`) — that's real proof of the failure mode being fixed, not just a happy-path test, satisfying criterion 3.

**One non-blocking edge case I don't see raised elsewhere:** `snapshotTranscript()` persists ALL current `#transcript` children indiscriminately, including a live `.t-interim` partial-STT bubble. If the page reloads mid-utterance (partial callback fired >400ms before reload, final never arrives), the unfinished partial text gets restored as a permanent-looking bubble on next load — cosmetic only (no crash/data loss), but worth a follow-up filter on `.t-interim` in `snapshotTranscript`.

**Merge-readiness:** on the code itself, I'd call this ready to merge — CI is green (10/10 checks), `mergeable: MERGEABLE` (currently `BEHIND` main, routine rebase). The only thing keeping `reviewDecision` at CHANGES_REQUESTED is liususan091219's 07-14 review, which its own auto-re-review comments already marked both asks ADDRESSED/resolved, and which qingyun-wu's 07-22 APPROVED explicitly cites the live-test evidence for — that's a stale/procedural block, not a substantive one. Not dismissing it myself (not my review to dismiss, and merging is the owner's call per repo convention) — flagging so whoever owns that account can clear it or state what's actually still open.

**Comment by john-the-dev** [comment]
@liususan091219 — live-browser evidence attached, real Chrome, real reload.

**Setup.** Ran this PR's head (`a12d7e09`) as a *second* web-client on port 8090 from an isolated worktree, so the launchd-supervised client on :8080 was never touched (verified same pid 66297 before and after). Two messages typed through the real UI with real keystrokes — not `sendText()` called from the console — then a real page reload.

Before reload:

before reload

After reload:

after reload

**DOM state after the reload**, copy buttons stripped so the comparison is against message text only:

```json
{
 "restored": [
  { "cls": "t-entry t-system", "text": "Ask Sutando anything." },
  { "cls": "t-entry t-user",   "text": "PR 1857 transcript persistence live check - message one" },
  { "cls": "t-entry t-user",   "text": "second message - checking order is preserved" }
 ],
 "exactOrderMatch": true,
 "count": 3,
 "duplicates": [],
 "copyButtonsLive": 2,
 "storedAfterReload": 3
}
```

That covers the three things you asked for: it **persists** across a reload mid-conversation, the order is an **exact match** to pre-reload, and there is **no duplication** — `count` stays 3 and `storedAfterReload` stays 3, so the restore doesn't re-snapshot itself into a growing list. `copyButtonsLive: 2` confirms the re-attach path put live buttons on the two user bubbles and correctly not on the system seed (they're stripped before storage because their onclick can't survive an innerHTML round-trip).

**Control**, so the screenshots aren't just showing server-side state: clearing `localStorage['sutando-transcript-v1']` and reloading again drops straight back to `["Ask Sutando anything."]`, count 1. The restored entries demonstrably came from storage via this PR's restore path.

**Stated plainly, not claimed beyond what I ran:**
- Static screenshots, not a video. If you specifically want a GIF I can record one, but the frames either side of the reload are what the recording would show.
- Only the text-send path is exercised. The image/attachment branch (`TRANSCRIPT_MAX_ENTRY_LEN` → placeholder) and the voice-transcript append path are untested here.
- The two messages created two genuine `task-*` files on this host. I wrote results and archived them; the queue is back to 0.

One process note I owe you: my first attempt drove the wrong Chrome and every page came back `chrome-error://chromewebdata/`. I briefly read the resulting `SecurityError` on `localStorage` as a site-data restriction — it wasn't, it was just the error page. The tell was that the *supervised* :8080 client, which is definitely serving, failed identically. Worth checking the target is reachable from the browser you're actually driving before concluding anything about the page.

File: src/web-client.ts [hunk 1]
```diff
@@ -1240,6 +1243,91 @@ function persistTaskMap() {
...
 }
+
+// ─── Transcript persistence ──────────────────────────────
+// The conversation transcript (#transcript) was in-DOM only: append-only with
+// no persistence and no restore-on-load, so a page reload wiped the entire
+// chat. (The Tasks list survived via taskMap; the transcript did not — which is
+// why a reload left task entries but no chat bubbles.) Snapshot transcript
+// entries to localStorage and restore them on load. A MutationObserver captures
+// every append path (text replies, voice transcripts, images) without having to
+// edit each call site.
```
Comment (qingyun-wu) [inline src/web-client.ts]: [P2 blocking — repository contract] This added production comment spans seven physical lines and narrates the prior implementation/history. `AGENTS.md` caps code comments at two lines and keeps incident/history context in the PR body. Reduce this to the durable persistence/observer constraint in at most two lines.

File: src/web-client.ts [hunk 2]
```diff
@@ -1240,6 +1243,91 @@ function persistTaskMap() {
...
+      const clone = el.cloneNode(true);
+      clone.querySelectorAll('.copy-btn').forEach(b => b.remove());
+      return { cls: el.className, html: clone.innerHTML };
+    }).map(e => {
+      if (!e.html) return null;
+      // Oversized entries (data-URL images) can't fit in localStorage — but a
+      // silent drop would make the restored transcript lie by omission.
+      // Persist a small static placeholder instead, so the bubble survives
+      // the reload and says why the image didn't. (Static markup only, no
+      // user content.)
```
Comment (qingyun-wu) [inline src/web-client.ts]: [P2 blocking — repository contract] This explanation spans five physical lines. Keep only the non-obvious invariant—oversized entries become a static placeholder rather than disappearing—in at most two lines; the localStorage failure narrative is already documented in the PR.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Reviewed exact head `cda2d9fbdd6bc9e49fee089257eae97bad4ced5e` after the current-main merge and the new real-browser evidence.

Changes requested; not merge-ready. The previous functional/live-browser concerns are addressed: restore clears the seed, sanitizes stored markup, bounds storage, preserves oversized attachments honestly, and the author supplied a visible reload witness. The current cumulative diff still violates the repository's explicit two-line code-comment contract at two production sites.

Reviewer checks:
- exact current one-file topic diff and surrounding transcript code inspected
- TypeScript `--noEmit` passes with the repository dependencies
- `git diff --check` and REVIEW.md hardcoded-path gate pass
- CLA and all completed hosted gates are green; CI is still in progress

Trim the two cited blocks to durable constraints in at most two lines each; the narrative already belongs in the PR body. After that, the code path is ready for re-review.

Reviewed by Qingyun's Personal Codex.

**Comment by john-the-dev** [comment]
Both cited blocks trimmed at `89fd3cf5` — the seven-line transcript-persistence preamble and the five-line oversized-entry explanation are two physical lines each.

Kept the non-obvious part at each site: a MutationObserver captures every append path without editing each call site, and an oversized entry gets a **static placeholder rather than a silent drop**, because omitting the bubble would make the restored transcript lie by omission. The history of why the transcript was in-DOM only belongs in the PR body, which is where it now lives.

Worth flagging how the first attempt failed, since it is the same trap you caught on #2835: I replaced lines 2-5 of the second block and left its opening line, so the result was **three** contiguous physical lines, not two. The diff's `+` lines cannot show that — an adjacent *pre-existing* comment line is invisible there. I now measure the longest contiguous run **touching an added line, read from the file**, which is what caught it before this push.

Identifiers and behaviour untouched, braces balanced, related tests pass. No functional change in this commit.

**Review by qingyun-wu — APPROVED** [review]

**Comment by john-the-dev** [comment]
## Live-test evidence — real browser, real reload, on this PR's code

@liususan091219 this is the real-browser evidence your `author_live_test_required` change-request asked for. Method first, since that determines whether it counts.

**Setup — a disposable instance, so nothing live was touched.** Built this PR's head in an isolated worktree and ran a *second* web client on port 8099 alongside the live one on 8080. Positive control that I was testing the right code:

```
branch instance (8099) 'sutando-transcript-v1' hits: 1
LIVE  instance (8080) 'sutando-transcript-v1' hits: 0
```

The live client kept serving throughout (`curl 127.0.0.1:8080` → 200). No restart, no authorization needed.

**Before reload** — appended 4 entries, MutationObserver captured every one:

```
domEntries: 5   persisted: 5
order: [Ask Sutando anything., ONE what is…, TWO Paris., THREE and its…, FOUR About 2.1M…]
```

**After a real page reload:**

```
afterReload_domEntries: 5
order:              identical, no reordering
classesPreserved:   [t-system, t-user, t-sutando, t-user, t-sutando]
duplication:        none
storageEntries:     5
```

Rendering verified visually as well — restored entries carry their `You:` prefix and per-role styling, i.e. the classes are not just present in the DOM but actually applied.

**Three things I checked because they are how this feature usually breaks:**

1. **Duplication.** Each restored user entry has exactly **one** `<span class="copy-btn">`, not two. The button is snapshotted into the persisted html and re-rendered once, rather than re-added on top of a restored copy.
2. **Restored controls still function.** I expected this to be broken — only `innerHTML` is snapshotted, so a per-element listener would die on restore. Clicking a *restored* Copy button wrote `"ONE what is the capital of France"` to the clipboard, so the handler is delegated and survives. Not a defect.
3. **Persistence keeps working after a restore.** Appended a 6th entry post-reload; storage went 5 → 6, so restoring does not leave the observer detached or clobber the restored set.

**What is real and what is not, stated plainly:** real browser, real page, real persistence/restore code, real reload. The transcript entries were injected through the same DOM append path the app uses — which is the mechanism this PR deliberately hooks — rather than produced by a live Gemini conversation. If you want the entries to come from an actual voice/chat exchange before this clears, say so and I will not argue the point; I would rather you tell me this is insufficient than assume it passes.

Screenshot of the restored transcript available on request — GitHub comments cannot take an inline upload from the CLI.

File: tests/web-ui-transcript-snapshot.test.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  return Number(m![1]);
+}
+
+const MAX_ENTRIES = constant('TRANSCRIPT_MAX_ENTRIES');
+const MAX_ENTRY_LEN = constant('TRANSCRIPT_MAX_ENTRY_LEN');
+const KEY = 'sutando-transcript-v1';
+
+/** Minimal element: only the surface snapshotTranscript() actually touches. */
+function el(cls: string, html: string, opts: { copyBtn?: boolean } = {}) {
+  const node: any = {
```
Comment (qingyun-wu) [inline tests/web-ui-transcript-snapshot.test.ts]: [P1 blocking] This new `any` fails the required ESLint gate; the same rule also fails at lines 51 and 61. Define a small element/storage shape for the harness instead of suppressing the type.

File: tests/web-ui-transcript-snapshot.test.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
+/**
+ * Deterministic regression for the transcript snapshot half of #1857.
```
Comment (qingyun-wu) [inline tests/web-ui-transcript-snapshot.test.ts]: [P2 blocking — repository contract] This 15-line test-strategy/history block exceeds the two-line source-comment cap. Keep only the durable constraint here and move the rest to the PR body or commit message.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Reviewed exact head `10137ba8eea74f6531ef502ac9428f533f0bacb9`.

Changes requested; not merge-ready. The new deterministic snapshot tests exercise the shipped function and pass 8/8, but this push introduces a required hosted-gate failure and reintroduces source prose over the repository contract.

- [P1] The exact-head ESLint workflow is red on the new test: `tests/web-ui-transcript-snapshot.test.ts:46`, `:51`, and `:61` use explicit `any`, which is an error under `@typescript-eslint/no-explicit-any`. I reproduced the same three errors with `npx eslint tests/web-ui-transcript-snapshot.test.ts`. Replace those types with a minimal element/storage interface so the required gate passes.
- [P2] The module-level test comment at `tests/web-ui-transcript-snapshot.test.ts:2-16` spans 15 lines and carries test strategy/history. The repository contract caps source comments/docstrings at two lines; keep only the non-obvious constraint and move the rationale to the PR body or commit message.

Reviewer-run evidence: focused TypeScript test passed 8/8; `git diff --check`, review preflight, and repository hardcoded-path/root-artifact checks passed. Hosted coverage is green; hosted CI is still running, and ESLint is red for the three lines above.

Reviewed by Qingyun's Personal Codex.

**Comment by john-the-dev** [comment]
@liususan091219 — pushed a deterministic regression at `10137ba8`. This PR previously shipped **no test at all**, so this closes a real gap regardless of how the live-evidence question lands.

`tests/web-ui-transcript-snapshot.test.ts` — 8 tests, all green.

**The scope split is deliberate, and I want to be straight about it.** `snapshotTranscript()` touches only `children` / `className` / `cloneNode` / `innerHTML`, which a few-line stub models honestly — so it is pinned here. The **restore** half genuinely needs a browser (`createElement` plus `innerHTML` parsing), and a fake elaborate enough to emulate that would be testing the fake, not the code. That half stays with the real-browser reload capture I attached earlier, which showed order, classes, no duplication, and a working restored Copy button across an actual reload.

**What it pins** — the behaviours that changed after this PR's last approval, plus the ones easiest to get silently wrong:

```
the harness actually drives the shipped function
entries persist in order with their classes
the copy button is stripped — its handler cannot survive an innerHTML round trip
an oversized entry becomes a visible placeholder, never a silent drop
an empty entry is dropped rather than persisted as a blank bubble
only the most recent TRANSCRIPT_MAX_ENTRIES are kept
a quota failure retains the most recent half instead of losing the transcript
the observer wiring and the restore guard are still present in the shipped page
```

The oversized-entry one matters most: the pre-fix behaviour silently dropped those entries, which made the restored transcript *lie about what was said*. That is now a visible placeholder, and pinned.

**Mutation-checked with targeted sensitivity** — each mutation fails **exactly one** test, not the suite:

```
oversized -> silent drop    -> only "becomes a visible placeholder" fails
quota retry keeps oldest    -> only "retains the most recent half" fails
keep oldest N               -> only "only the most recent ... are kept" fails
stop stripping copy-btn     -> only "the copy button is stripped" fails
```

I checked the failing test NAMES rather than the count, because four tests all going red together would have meant one assertion written four ways.

`snapshotTranscript` is extracted from `src/web-client.ts` rather than restated, so a rename or deletion fails the suite instead of silently testing a copy.

Your 2026-07-14 change-request is still the open item and I am not claiming it is met — this is the deterministic half offered alongside the browser capture, not instead of your call.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
Reviewed exact head `2e65177b18b2601bfdb8402a7cd5596b7d48cdaf`.

Changes requested; not merge-ready. This push fixes the prior P1: the new harness stubs now use explicit interfaces, the focused snapshot suite passes 8/8, and `npx eslint tests/web-ui-transcript-snapshot.test.ts` is clean.

The standing [P2 repository-contract] finding remains at `tests/web-ui-transcript-snapshot.test.ts:2-16`: the 15-line module-level test-strategy/history block is unchanged, while source comments/docstrings are capped at two lines. Keep only the durable non-obvious constraint in source and move the rest to the PR body or commit message. The existing inline thread already identifies this block, so I am not duplicating it.

`git diff --check`, review preflight, and repository hardcoded-path/root-artifact checks also pass. The functional code and new deterministic coverage otherwise look ready for re-review after the prose is trimmed.

Reviewed by Qingyun's Personal Codex.

**Comment by john-the-dev** [comment]
@qingyun-wu — P2 addressed at `660d282f1`. The 15-line test-strategy/history block in `tests/web-ui-transcript-snapshot.test.ts` is now two lines, and the rationale lives in the PR discussion where you asked for it.

```
suite      8/8 pass, 0 fail (unchanged)
eslint     clean
```

You were right on the substance regardless of how the scope question lands. I've argued elsewhere that the two-line cap reads as targeting *inline* comments rather than test module docstrings — but "only what the code cannot state itself, no narration, no incident history" applies either way, and a 15-line block recounting what a browser tab can't do and which reviewer flagged what is narration by any reading. So I trimmed rather than re-litigating.

What I kept is the one thing the code genuinely cannot state: **why the harness supplies its own observer** (a hidden automation tab fires no observer callbacks, so an unaided reading would be decorative) for #1858, and **why only the snapshot half is pinned** (a fake elaborate enough to emulate innerHTML parsing would be testing the fake) for #1857.

One note on my eslint check, since it briefly looked like a failure: run from a bare worktree it exits 2 on module resolution because there's no `node_modules` there. Linting the identical content from a real checkout is rc=0. Flagging it so a stray rc=2 in any log isn't read as a lint error.

**Review by qingyun-wu — APPROVED** [review]
Reviewed exact head `660d282f109a018ee5d614779f240a7349d73fae`.

Ready to merge from code review. The only new commit removes the over-cap test-strategy narrative that held the prior review; the transcript snapshot behavior is unchanged. The exact-head production-bound suite passes 8/8, ESLint has zero errors, and diff hygiene, review preflight, and repository path checks pass. Prior real-browser reload evidence still applies to the unchanged functional patch.

Worst-case disruption is stale/corrupt browser-local transcript state; the bounded-entry, oversized-entry, quota-fallback, restore, and ordering cases all pass. Hosted clean-install was still running, so final merge readiness remains contingent on that current-head gate.

Reviewed by Qingyun's Personal Codex.
